rust/worker/src/catch_panic.rs

```diff
@@ -0,0 +1,75 @@
+use std::{
+    panic::AssertUnwindSafe,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::FutureExt;
+use hyper::body::Body;
+use tonic::body::BoxBody;
+use tower::{Layer, Service};
+
+#[derive(Debug, Clone, Default)]
+pub struct CatchPanicLayer;
+
+impl<S> Layer<S> for CatchPanicLayer {
+    type Service = CatchPanicMiddleware<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        CatchPanicMiddleware { inner: service }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CatchPanicMiddleware<S> {
+    inner: S,
+}
+
+type BoxFuture<'a, T> = Pin<Box<dyn std::future::Future<Output = T> + Send + 'a>>;
+
+impl<S> Service<hyper::Request<Body>> for CatchPanicMiddleware<S>
+where
+    S: Service<hyper::Request<Body>, Response = hyper::Response<BoxBody>> + Clone + Send + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: hyper::Request<Body>) -> Self::Future {
+        // This is necessary because tonic internally uses `tower::buffer::Buffer`.
+        // See https://github.com/tower-rs/tower/issues/547#issuecomment-767629149
+        // for details on why this is necessary.
+        // (We could also probably avoid a clone using an approach similar to https://docs.rs/tower-http/latest/tower_http/catch_panic, but wrangling the types between hyper/tower/tonic is very tricky.)
```

Comment:
spent a while trying to get this approach to work, didn't seem worth it in the short term
if we profile and discover this clone is very expensive you know who to call ;)